analog_io_modules/seven_sensor_rain_gauge/firmware.lua

```diff
@@ -0,0 +1,206 @@
+DATA_BITS = 'data_bits'
+SERIAL_PORT = 'serial_port'
+
+
+-- Initiate device firmware. Called at the end of the file.
+
+function main()
+  config.init({
+    [BAUDRATE] = { type = 'number', required = true, default = 9600 },
+    [STOP_BITS] = { type = 'number', required = true, default = 1 },
+    [PARITY] = { type = 'string', required = true, default = 'N' },
+    [ADDRESS] = { type = 'number', required = true, default = 1 },
+    [DATA_BITS] = { type = 'number', required = true, default = 8 },
+    [SERIAL_PORT]= {type = 'string', required = true, default = '/dev/ttyS0'}
+  })
+  scheduler.add(30000, send_properties)
+  scheduler.add(1000, send_telemetry)
+end
+
+function send_properties()
+  local properties = {}
+
+  properties.vendor = VENDOR
+  properties.model = MODEL
+
+  enapter.send_properties(properties)
+end
+
+function send_telemetry()
+
+  local device, err = connect_device()
+  if not device then
+    if err == 'cannot_read_config' then
+      enapter.send_telemetry({ status = 'error', alerts = { 'cannot_read_config' } })
+    elseif err == 'not_configured' then
+      enapter.send_telemetry({ status = 'ok', alerts = { 'not_configured' } })
+    end
+    return
+  end
+
+  local telemetry = {}
+  local alerts = {}
+  local status = 'ok'
+
+  local data, result = device:read_holdings(address, 40020, 16, 1000)
+    if data then
+      telemetry['model'] = toSunSpecStr(data)
+    else
+      enapter.log('Register 40020 reading failed: ' .. modbus.err_to_str(result), 'error')
+      status = 'read_error'
+    end
+
+  local data, result = device:read_holdings(address, 40052, 16, 1000)
+    if data then
+      telemetry['serial number'] = toSunSpecStr(data)
+    else
+      enapter.log('Register 40052 reading failed: ' .. modbus.err_to_str(result), 'error')
+      status = 'read_error'
+  end
+
+  local data, err = device:read_inputs(address, 30022, 1, 1000)
+  if data then
+    telemetry['Rain Gauge(Hour)']= table.unpack(data) / 10.0
+  else
+    enapter.log('Register 30022 reading failed: no data'.. modbus.err_to_str(err), 'error')
+    status = 'read_error'
+  end
+
+  local data, err = device:read_inputs(address, 30028, 1, 1000)
+  if data then
+    telemetry['Rain Gauge(Min)']= table.unpack(data) / 10.0
+  else
+    enapter.log('Register 30022 reading failed: no data'.. modbus.err_to_str(err), 'error')
+    status = 'read_error'
+  end
+
+  local data, err = device:read_inputs(address, 30029, 1, 1000)
+  if data then
+    telemetry['Rain Gauge(Sec)']= table.unpack(data) / 10.0
+  else
+    enapter.log('Register 30022 reading failed: no data'.. modbus.err_to_str(err), 'error')
+    status = 'read_error'
+  end
+
+  telemetry.alerts = alerts
+  telemetry.status = status
+  enapter.send_telemetry(telemetry)
+end
+
+function toSunSpecStr(registers)
+  local str = ''
+  for _, reg in pairs(registers) do
+    local msb = reg >> 8
+    local lsb = reg & 0xFF
+    str = str .. string.char(lsb) .. string.char(msb)
+  end
```

Comment:
you can wrap this for loop into [pcall ](https://www.lua.org/pil/8.4.html)

Reply:
like this?

Reply:
not really. I meant the whole for loop. if this function will receive nil instead of table, then your code will panic.